### PR TITLE
[Asyncio] Increase recursion limit manually

### PR DESCRIPTION
**Pull request by simon-mo**
This PR address the issue that asyncio actor segfaulting in python 3.7. 

Previously, we used generators to make sure python doesn't mistakenly count our executor stack as part of their recursion limit. However, this approach will cause segfault in python3.7. Instead of relying on python internal implementation, we will manually double the recursion limit when we are close to the limit so Python won't error. The doubling only happens before the task execute. This means if user have an infinite recursive call, Python can still catch it and error with RecusionLimitExceeded.

To get the current recursion depth, we have to ping cpython internal API, there's no way around that.

### CODE REVIEW

File: python/ray/_raylet.pyx [hunk 1]
```diff
@@ -385,13 +393,18 @@ cdef execute_task(
...
 
-            if PY3 and core_worker.current_actor_is_asyncio():
+            if core_worker.current_actor_is_asyncio():
+                # Increase recursion limit if necessary. In asyncio mode,
+                # we have many parallel callstacks (represented in fibers)
+                # that's suspended for execution. Python interpreter will
+                # mistakenly count each callstack towards recusion limit.
+                # We don't need to worry about stackoverflow here because
+                # the max number of callstacks is limited in direct actor
+                # transport with max_concurreny flag.
```
Comment (edoakes) [inline python/ray/_raylet.pyx]:
```suggestion
                # transport with max_concurrency flag.
```

File: python/ray/_raylet.pyx [hunk 2]
```diff
@@ -207,6 +199,21 @@ def compute_task_id(ObjectID object_id):
     return TaskID(object_id.native().TaskId().Binary())
 
 
+cdef increase_recursion_limit():
```
Comment (edoakes) [inline python/ray/_raylet.pyx]: Maybe we can just initially set the recursion limit to `20 * max_concurrency`? Might be a bit simpler

File: python/ray/_raylet.pyx [hunk 3]
```diff
@@ -412,14 +425,14 @@ cdef execute_task(
...
                 with nogil:
                     (core_worker.core_worker.get()
                         .YieldCurrentFiber(fiber_event))
 
-                yield future.result()
+                result = future.result()
+                return result
 
-            yield function(actor, *arguments, **kwarguments)
+            return function(actor, *arguments, **kwarguments)
```
Comment (edoakes) [inline python/ray/_raylet.pyx]: Why the change from yield to return here?

File: python/ray/_raylet.pyx [hunk 4]
```diff
@@ -412,14 +425,14 @@ cdef execute_task(
...
 
                 future.add_done_callback(callback)
-
                 with nogil:
                     (core_worker.core_worker.get()
                         .YieldCurrentFiber(fiber_event))
 
-                yield future.result()
+                result = future.result()
+                return result
```
Comment (edoakes) [inline python/ray/_raylet.pyx]: should just return on previous line

**Review by edoakes — COMMENTED** [review]
This looks like it wasn't fun to debug! Can you also add a test for it?

File: python/ray/_raylet.pyx (hunk 2 above)
Comment (simon-mo) [reply python/ray/_raylet.pyx]: That won't quite work because we don't know how many stack is there per actor call. The doubling approach is less brittle

File: python/ray/_raylet.pyx (hunk 3 above)
Comment (simon-mo) [reply python/ray/_raylet.pyx]: Explained in comments above. `yield` mark the function generator, we don't want that.

**Comment by simon-mo** [comment]
There's a test for it in test_asyncio::test_asncio_actor_high_concurrency where we go above the recursion limit. The real issue is that we didn't test against python 3.7, which we should consider running for release.

**Review by edoakes — APPROVED** [review]
Thanks for clarifications - LGTM.
